distributedlog-io/dlfs/src/main/java/org/apache/distributedlog/fs/DLFileSystem.java

```diff
@@ -0,0 +1,330 @@
+        }
+        log.info("Initializing the filesystem at {}", name);
+        // initialize the namespace
+        this.namespace = NamespaceBuilder.newBuilder()
+                .clientId("dlfs-client-" + InetAddress.getLocalHost().getHostName())
+                .conf(dlConf)
+                .regionId(DistributedLogConstants.LOCAL_REGION_ID)
+                .uri(name)
+                .build();
+        log.info("Initialized the filesystem at {}", name);
+    }
+
+    @Override
+    public void close() throws IOException {
+        // clean up the resource
+        namespace.close();
+        super.close();
+    }
+
+    //
+    // Util Functions
+    //
+
+    private Path makeAbsolute(Path f) {
+        if (f.isAbsolute()) {
+            return f;
+        } else {
+            return new Path(workingDir, f);
+        }
+    }
+
+    private String getStreamName(Path relativePath) {
+        return makeAbsolute(relativePath).toUri().getPath().substring(1);
+    }
+
+    //
+    // Home & Working Directory
+    //
+
+    @Override
+    public Path getHomeDirectory() {
+        return this.makeQualified(new Path(System.getProperty("user.home", "")));
+    }
+
+    protected Path getInitialWorkingDirectory() {
+        return this.makeQualified(new Path(System.getProperty("user.dir", "")));
+    }
+
+    @Override
+    public void setWorkingDirectory(Path path) {
+        workingDir = makeAbsolute(path);
+        checkPath(workingDir);
+    }
+
+    @Override
+    public Path getWorkingDirectory() {
+        return workingDir;
+    }
+
+
+    @Override
+    public FSDataInputStream open(Path path, int bufferSize)
+            throws IOException {
+        try {
+            DistributedLogManager dlm = namespace.openLog(getStreamName(path));
+            LogReader reader;
+            try {
+                reader = dlm.openLogReader(DLSN.InitialDLSN);
+            } catch (LogNotFoundException lnfe) {
+                throw new FileNotFoundException(path.toString());
+            } catch (LogEmptyException lee) {
+                throw new FileNotFoundException(path.toString());
+            }
+            return new FSDataInputStream(
+                new BufferedFSInputStream(
+                    new DLInputStream(dlm, reader, 0L),
+                    bufferSize));
+        } catch (LogNotFoundException e) {
+            throw new FileNotFoundException(path.toString());
+        }
+    }
+
+    @Override
+    public FSDataOutputStream create(Path path,
+                                     FsPermission fsPermission,
+                                     boolean overwrite,
+                                     int bufferSize,
+                                     short replication,
+                                     long blockSize,
+                                     Progressable progressable) throws IOException {
+        // for overwrite, delete the existing file first.
+        if (overwrite) {
+            delete(path, false);
```

Comment:
what happens in the case that create fails but you managed to delete the file?

Reply:
current approach doesn't guarantee any atomicity. this can be improved if dlog api supports opening a log stream in overwrite mode.

Reply:
Create an issue for this - https://github.com/apache/distributedlog/issues/231